src/validation.cpp

```diff
@@ -3222,6 +3253,16 @@ static bool FindBlockPos(FlatFilePos &pos, unsigned int nAddSize, unsigned int n
 
     if (!fKnown) {
         while (vinfoBlockFile[nFile].nSize + nAddSize >= MAX_BLOCKFILE_SIZE) {
```

Comment:
I find this `while` extremely confusing - it looks to me that it will never iterate more than once. We need a block size of more than 128MB (blk file size limit) to get this execute 2 or more times, so that one block spans multiple blk files (!?).

Maybe out of the scope of this PR, but I think we should `s/while/if/`. Is my understanding correct? If not then we will end up finalizing an empty undo file with this patch on the second and subsequent iterations.

Reply:
I dug into this. Apparently, the system is supposed to support multiple files *per block*. This was a change made 8 years ago in 5382bcf8cd23c36a435c29080770a79b5e28af42 by @sipa. It doesn't look like it is actually working, though, but I could be wrong.

Reply:
@kallewoof No, that commit adds support for multiple *blocks per file*, which is definitely used. There has never been support for multiple files per block.

Reply:
Oh, yeah you're right. But this while loop makes no sense to me still.

Reply:
I think it's just iterating to find the first block file with space for the new block.

Reply:
Okay, so assuming we one day decided to stop writing block files sequentially, we may have a situation where we try to `find block pos` for `nFile = 2007` where the last block file in existence is actually `nFile = 2010` or something `>2008`, in which case this while loop becomes necessary.

I don't really see that ever happening though, and it feels like someone got "block file writing order" (sequentially; we never write into older block files or into the middle of a file in between other stuff) and "connect order" (non-sequentially, as blocks are not stored in height order).

Reply:
Ok, for the purposes of this PR we just need to agree that this `while` loop will never iterate more than once (if it is possible that it iterates 2 or more times then I am wrong and this PR will have to be adjusted).

Observation: before the loop condition is evaluated for the first time `nFile` equals `nLastBlockFile`. Ie we start the search at the end.

So two things can happen:
1. We find space in the last file - the loop condition evaluates to `false` and the loop iterates 0 times
2. There is no space in the last file, the loop condition evaluates to `true` and the loop body "creates" a new empty block file (not yet on disk, but this is not relevant). The new empty block file has `nSize == 0` and so the next loop condition will for sure evaluate to `false` because `nSize + addedBlockSize` cannot be `>=` than 128 MiB.

> I think it's just iterating to find the first block file with space for the new block.

Yes, and given that it always starts the search at the last file it will never iterate more than once, as per above.

Reply:
I think in the original version of this code the loop was there so that after a restart a good file number would be found (even if it always reset to 0). I don't know if that's still needed.